### PR TITLE
bugfix/2020_2021_input-text-currency

### LINKED ISSUE #2020: Input-Text: no currency formatting on re-render
Remo reported:

1. set currency = CHF
2. type and blur -> everything ok
3. trigger re-render (in react show another component and go back)
4. the input-text value is just the number without currency format

### LINKED ISSUE #2021: Input-Text: if currency set the invalid flag is ignored
Remo reported:

[reproducer](https://patterns.axa.ch/?path=/story/components-input-text--input-text&knob-label*=&knob-name*=&knob-refid=&knob-placeholder=&knob-value=3&knob-currency=chf&knob-error=fehler&knob-info=&knob-checkmark=&knob-disabled=&knob-required=&knob-invalid=true&knob-type=text&knob-maxlength=50&knob-counter=Still ##counter## characters left&knob-counterMax=Over character limit!&knob-pattern=&knob-inputmode=&knob-autofocus=)

active/deactivate the "invalid" knob.... nothing happens

### PR DESCRIPTION
**Pull request by raphael-suter**
Fixes #2020 
Fixes #2021

# Merge Checklist:
- [x] Code selfreview has been performed ([Best Practices](https://github.com/axa-ch/patterns-library/blob/develop/CONTRIBUTION.md#best-practices)).
- [x] Tests are sufficient.
- [x] Modifications still work in IE.
- [x] Documentation is up to date.
- [x] Changelog is up to date.
- [x] Typescript typings for properties are up to date.
- [x] Designers approved changes or no approval needed.
- [x] Dependencies to other components still work.

### CODE REVIEW

File: src/components/10-atoms/input-text/index.js [hunk 1]
```diff
@@ -272,23 +280,23 @@ class AXAInputText extends AXAPopupMixin(NoShadowDOM) {
...
 
   firstUpdated() {
-    const { defaultValue, isReact, value } = this;
+    const { defaultValue, isReact } = this;
     this._setNativeInput();
 
-    if (isReact) {
-      this.nativeInput.value = defaultValue || value;
+    if (isReact && defaultValue) {
+      this.nativeInput.value = this._formatCurrency(defaultValue);
```
Comment (danyball) [inline src/components/10-atoms/input-text/index.js:299]: You has added `defaultValue` to the if. And removed value. Isnt it possible that defaultValue is false and value is truthy?

File: src/components/10-atoms/input-text/ui.test.js [hunk 2]
```diff
@@ -240,6 +240,33 @@ test('should format value of controlled component', async t => {
...
+  }).find(CLASS);
+
+  await t.expect($axaInputElement.value).eql('CHF 45.00');
+});
+
+fixture('Input text - currency validation and property invalid').page(
+  `${host}/iframe.html?id=components-input-text--input-text&knob-label%2A=&knob-name%2A=&knob-refid=&knob-placeholder=&knob-value=2&knob-currency=chf&knob-error=&knob-info=&knob-type=text&knob-maxlength=50&knob-counter=Still%20%23%23counter%23%23%20characters%20left&knob-counterMax=Over%20character%20limit%21&knob-pattern=&knob-inputmode=&knob-invalid=true&viewMode=story`
+);
+
+test('should display error when invalid is set', async t => {
```
Comment (danyball) [inline src/components/10-atoms/input-text/ui.test.js:259]: Maybe add another test to check the visibility of the error TEXT if it is set.

File: src/components/10-atoms/input-text/index.js (hunk 1 above)
Comment (raphael-suter) [reply src/components/10-atoms/input-text/index.js:299]: Yes thats true but `value` is set during `render` and therefore there is no need to set it again.

File: src/components/10-atoms/input-text/ui.test.js (hunk 2 above)
Comment (raphael-suter) [reply src/components/10-atoms/input-text/ui.test.js:259]: 👍 added some lines

File: src/components/10-atoms/input-text/README.md [hunk 3]
```diff
@@ -164,7 +164,7 @@ The Boolean attribute autofocus, when true, sets keyboard focus on the underlyin
 
 You can set the `currency` attribute to a valid [ISO 4217 currency code](https://en.wikipedia.org/wiki/ISO_4217) string. When set, &lt;input-text&gt; formats the input whenever the `onBlur` event of its internal input element fires, and also upon setting `value` programmatically. Note that the locale "style" remains fixed to `de-CH`.
 
-Currency formatting is only active when attribute `type` is set to `text`. If the inputted text can not be formatted the attribute `invalid` switches to `true`. The Formatter do not accept values which are only letters ("onehundret") or has more than 1 decimal separators ("12.30.1").
+Currency formatting is only active when attribute `type` is set to `text`. The Formatter do not accept values which are only letters ("onehundret") or has more than 1 decimal separators ("12.30.1").
```
Comment (raphaellueckl) [inline src/components/10-atoms/input-text/README.md]:
```
The formatter does not accept values, which are either only letters ("onehundret") or have more than one decimal separator ("12.30.1").
```

Comment (01231) [reply src/components/10-atoms/input-text/README.md]: google says `one hundred` is the correct spelling not `onehundret`

**Review by 01231 — CHANGES_REQUESTED** [review]
Can't type in input-text demo - on prod this works :(
https://axa-ch.github.io/plib-feature/bugfix/2020_2021_input-file-currency/?path=/story/examples-input-text-react--story

File: package.json [hunk 4]
```diff
@@ -22,7 +22,7 @@
     "test-jest-coverage": "jest --env=jsdom --coverage",
     "test-jest-watch": "jest --env=jsdom --watch -o",
     "test-ui": "TEST_HOST_STORYBOOK_URL=http://localhost:6006 testcafe chrome ./**/ui.test.js",
-    "test-ui-hl": "TEST_HOST_STORYBOOK_URL=http://localhost:6006 testcafe 'chrome:headless' ./**/ui.test.js",
+    "test-ui-hl": "TEST_HOST_STORYBOOK_URL=http://localhost:6006 testcafe 'chrome:headless' ./**/input-text/ui.test.js",
```
Comment (danyball) [inline package.json]: Pls remove that

**Review by danyball — CHANGES_REQUESTED** [review]
Pls have a look at noahs comment.

**Review by 01231 — DISMISSED** [review]

File: src/components/10-atoms/input-text/README.md [hunk 5]
```diff
@@ -164,7 +164,7 @@ The Boolean attribute autofocus, when true, sets keyboard focus on the underlyin
 
 You can set the `currency` attribute to a valid [ISO 4217 currency code](https://en.wikipedia.org/wiki/ISO_4217) string. When set, &lt;input-text&gt; formats the input whenever the `onBlur` event of its internal input element fires, and also upon setting `value` programmatically. Note that the locale "style" remains fixed to `de-CH`.
 
-Currency formatting is only active when attribute `type` is set to `text`. If the inputted text can not be formatted the attribute `invalid` switches to `true`. The Formatter do not accept values which are only letters ("onehundret") or has more than 1 decimal separators ("12.30.1").
+Currency formatting is only active when attribute `type` is set to `text`. The formatter does not accept values, which are either only letters ("one hundred") or have more than one decimal separator ("12.30.1").
```
Comment (markus-walther) [inline src/components/10-atoms/input-text/README.md]: drop the comma

File: src/components/10-atoms/input-text/README.md [hunk 6]
```diff
@@ -174,7 +174,8 @@ Please be mindful of using currency formatting in combination with `maxLength`.
 
 The function-valued property `onChange` can be used as a callback prop for React and other frameworks.
 
-Its only argument is the original `change` event from the element's underlying native &lt;input&gt; field.
+Its first argument is the original `change` event from the element's underlying native &lt;input&gt; field.
+The second argument is the internal `invalidFormat` property, which is `true` when the user's input can't be formatted.
```
Comment (markus-walther) [inline src/components/10-atoms/input-text/README.md]: Suggestion: The second argument is a Boolean flag `invalidformat` which is true if, and only if, the user's input cannot be formatted as a currency.

(You don't want to tell developers about an undocumented internal property...)

File: src/components/10-atoms/input-text/index.js [hunk 7]
```diff
@@ -210,7 +216,8 @@ class AXAInputText extends AXAPopupMixin(NoShadowDOM) {
   };
 
   handleInput = ev => {
-    this.onChange(ev);
+    this._formatCurrency(ev.target.value);
```
Comment (markus-walther) [inline src/components/10-atoms/input-text/index.js:220]: add comment: // sets this.invalidFormat

**Review by markus-walther — CHANGES_REQUESTED** [review]
Great work. Just 3 small nits to do with language and commenting.

**Review by markus-walther — APPROVED** [review]
LGTM

**Pull request merged**